src/elf/mod.rs

```diff
@@ -131,23 +131,56 @@ if_sylvan! {
     }
 
     impl<'a> Elf<'a> {
-        /// Try to iterate the notes; returns `None` if there aren't any notes in this binary
-        pub fn iter_notes(&self, data: &'a [u8]) -> Option<note::NoteIterator<'a>> {
+        /// Try to iterate notes in PT_NOTE program headers; returns `None` if there aren't any note headers in this binary
+        pub fn iter_note_headers(&self, data: &'a [u8]) -> Option<note::NoteIterator<'a>> {
+            let mut iters = vec![];
             for phdr in &self.program_headers {
                 if phdr.p_type == program_header::PT_NOTE {
                     let offset = phdr.p_offset as usize;
                     let alignment = phdr.p_align as usize;
-                    return Some(
-                        note::NoteIterator {
-                            data,
-                            offset,
-                            size: offset + phdr.p_filesz as usize,
-                            ctx: (alignment, self.ctx)
-                        }
-                    )
+
+                    iters.push(note::NoteDataIterator {
```

Comment:
We can make this not allocate, but it would require another iterator struct which basically implements this logic, but I'm not sure its worth it at all. Alternatively, when `impl Trait` comes will be easy fix

Reply:
I started that way but then the solution seemed way to complex for what it was trying to do. `impl Trait` will really make iterators much easier. Until then, I'd propose to leave it that way. Since all section and program headers are already parsed at that point, the only optimization would be to save a single (in **very** rare cases more) allocation.